File: LeetCode_Algorithm/876/876.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from typing import Optional

def makeTree(head, idx):
    rootNode = ListNode(head[idx])
    if idx+1 < len(head): rootNode.next = makeTree(head, idx+1)
    return rootNode

class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        rootNode = makeTree(head, 0)
        # # sol1 O(n), O(1)  36~52ms 73~27% 
        # slow = rootNode
        # fast = rootNode
        # if slow.next is None: return slow.val

        # while fast is not None and fast.next is not None:
        #     slow = slow.next
        #     # if fast is not None and fast.next is not None:
        #     #     fast = fast.next.next
        #     # elif fast is not None:
        #     #     fast = fast.next

        #     # 어차피 루프를 돌기 위해서는 next가 존재해야 한다고 했으므로
        #     fast = fast.next.next

        # return slow.val


        # sol2 O(n), O(n)
        arr = [rootNode]
        while arr[-1].next:
            arr.append(arr[-1].next)
        return arr[len(arr)//2].val



            

        



print(Solution().middleNode([1,2,3,4]))

File: LeetCode_Algorithm/876/876.js
function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : val)
}
var makeTree = function(head, idx) {
  var rootNode = new ListNode(head[idx])
  if (idx+1 < head.length) {rootNode.next = makeTree(head, idx+1)}
  return rootNode
}
var middleNode = function(head) {
  var rootNode = makeTree(head, 0)
  // sol1 O(n), O(1) 70~87ms 66~37%
  //  if (rootNode.next == null) {return rootNode}
  
  //  var slow = rootNode; var fast = rootNode;
  //  while (fast != null && fast.next != null) {
  //    console.log(slow.val, fast.val)
  //    slow = slow.next
  //    fast = fast.next.next
  //  }
  //  return slow.val


  // sol2 O(n), O(n) 83~96ms 43~21%
  arr = [rootNode]
  while (arr[arr.length-1].next != null) {
    arr.push(arr[arr.length-1].next)
  }
  return arr[parseInt(arr.length/2)]
  
}
console.log(middleNode([1,2,3,4,5,6]))